dev/bus/export/dev/serial.hpp

```diff
@@ -0,0 +1,217 @@
+    static err recv_buf(uint8_t *buf, size_t &sz);
+
+    //! Sends byte to a serial device.
+    //! \pre Driver is initialized.
+    //! \details It may not block if internal buffering is applied.
+    //! It may also block if platform device is not yet ready to
+    //! send data.
+    //! \param[in] byte Byte to send.
+    //! \return Operation status.
+    static err send_byte(uint8_t byte);
+
+
+    //! Sends buffer to a serial stream.
+    //! \pre Driver is initialized.
+    //! \details It may not block if internal buffering is applied.
+    //! It may also block if platform device is not yet ready to
+    //! send data.
+    //! \param[in]      buf Buffer to send.
+    //! \param[in,out]  sz  Size of buffer. Will be updated with
+    //!                     amount of bytes sent In case of error,
+    //!                     size will be updated with amount of bytes
+    //!                     successfully sent before error occur.
+    //! \return Operation status.
+    static err send_buf(const uint8_t *buf, size_t &sz);
+
+private:
+    static void bus_handler(bus_channel ch, bus_event type, size_t total);
+
+    static err try_start_xfer();
+
+    static uint8_t m_rx_buffer[buf_size];
+    static uint8_t m_tx_buffer[buf_size];
+
+    static binary_semaphore m_rx_new_xfer_allowed;
+
+    static std::atomic_size_t m_rx_write_iter; //!< Used for accepting data from the bus
+    static std::atomic_size_t m_rx_read_iter; //!< Used for returning data to the user
+
+    static bool m_is_inited;
+};
+
+template <class PBus, size_t buf_size>
+uint8_t serial<PBus, buf_size>::m_rx_buffer[buf_size];
+
+template <class PBus, size_t buf_size>
+uint8_t serial<PBus, buf_size>::m_tx_buffer[buf_size];
+
+template <class PBus, size_t buf_size>
+std::atomic_size_t serial<PBus, buf_size>::m_rx_write_iter;
+
+template <class PBus, size_t buf_size>
+std::atomic_size_t serial<PBus, buf_size>::m_rx_read_iter;
+
+template <class PBus, size_t buf_size>
+bool serial<PBus, buf_size>::m_is_inited;
+
+template <class PBus, size_t buf_size>
+binary_semaphore serial<PBus, buf_size>::m_rx_new_xfer_allowed;
+
+template <class PBus, size_t buf_size>
+err serial<PBus, buf_size>::init()
+{
+    ecl_assert(m_is_inited == false);
+    auto result = PBus::init();
+    if (is_error(result)) {
+        return result;
+    }
+    PBus::set_handler(bus_handler);
+    result = PBus::enable_listen_mode();
+    if (is_error(result)) {
+        return result;
+    }
+    m_rx_new_xfer_allowed.signal();
+    PBus::set_rx(m_rx_buffer, buffer_size);
+    result = PBus::do_xfer();
+    if (is_ok(result)) {
+        m_is_inited = true;
+    }
+    return result;
+}
+
+template <class PBus, size_t buf_size>
+void serial<PBus, buf_size>::bus_handler(bus_channel ch,
+        bus_event type, size_t total)
+{
+    if (ch == bus_channel::rx) {
+        ecl_assert(type == bus_event::tc);
+        m_rx_write_iter++;
+        ecl_assert(m_rx_write_iter == total);
+
```

Comment:
Overflow flag must be raised somewhere here. Proposed condition:
```

# Last time buffer was full, and no one has read any byte from it.
if buffer_is_full() and read_iter == write_iter:
   set_overflow()
   # Advance read_iter, so overflow condition will stay there until 
   # somebody will read data. And read iterator will point to the start
   # of the old, overflowed chunk of data. This has to be discussed.
   read_iter++;

# Checks that buffer is about to be full
if read_iter > 1 && write_iter == read_iter - 1: 
   set_buffer_full()

write_iter++
```

It means that . On the next ISR call if no read